include/seastar/core/reactor.hh

```diff
@@ -31,6 +31,7 @@
 #include <seastar/core/file.hh>
 #include <seastar/core/future.hh>
 #include <seastar/core/idle_cpu_handler.hh>
+#include <seastar/core/internal/io_desc.hh>
```

Comment:
So you depopulated internal header at the cost of bloating core public one. That's not very nice, reactor.hh should be put on a diet instead

Reply:
No, that's not the case. Please let me explain and attach some references.

Firstly, `reactor.hh` was already bloated and it was easy to break in unexpected way. I will start from the part related to `breaking in unexpected way`.

Let's assume that you are a programmer and you open `pollable_fd.hh`. You see, that nothing in this file needs `kernel_completion` that is the only type defined in `io_desc.hh`. This means that the inclusion of that header is redundant. Normally, because of programming "scout rule" you want to leave the file cleaner when you leave it, so you remove the unneeded include to remove the dependency that is in fact redundant.

What happens after the removal? The project stops to compile, because `reactor.hh` was not self-contained [[1]](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#sf11-header-files-should-be-self-contained). According to cppcoreguidlines it is a good rule of thumb to make all headers self-contained.

In `reactor.hh` there was a forward declaration of `kernel_completion`  and then we tried to derive a type from that forward declaration. This is illegal [[2]](https://godbolt.org/z/7aK8aGsW5).

What's more important, `reactor.hh` was already bloated, because of the way how `#include` directive works. It is just a simple copy-paste tool. Therefore, if anybody included `reactor.hh` he/she implicitly received `io_desc.hh`, because `reactor.hh` includes `pollable_fd.hh` and it includes `io_desc.hh`.

What is wrong is the fact, that any file that includes `pollable_fd.hh` implicitly includes `io_desc.hh` header, that is not needed by `pollable_fd.hh`. If it is not needed also by the client of `pollable_fd.hh` then we created unneeded dependency.

Is this explanation enough? Honestly, I do not understand the meaning of `the cost of bloating core public one` -- the public one is already bloated on master and is easy to break due to invalid usage of forward declaration and it compiles just by luck and implicit include of `io_desc.hh` via `pollable_fd.hh` that in fact should not include it.

Reply:
I see. OK, this makes sense. Please update the description to mention that reactor.hh pulled kernel_completion via pollable_fd.hh. And this PR cannot now be merged due to conflict, so please rebase as well

Reply:
Sure, fixed. Thanks :+1: